From: Felix Fietkau <nbd@nbd.name>
Date: Wed, 1 May 2024 18:40:19 +0200
Subject: [PATCH] ubus: fix uc_ubus_have_uloop for eloop+uloop combination

When uloop has been integrated with eloop (in hostapd/wpa_supplicant),
uloop_cancelling will return false, since uloop_run is not being called.
This leads to ubus.defer() calling uloop_run in a context where it can
prevent the other event loop from running.

Fix this by detecting event loop integration via uloop_fd_set_cb being set

Signed-off-by: Felix Fietkau <nbd@nbd.name>
---

Index: ucode-2023-06-06-c7d84aae/lib/ubus.c
===================================================================
--- ucode-2023-06-06-c7d84aae.orig/lib/ubus.c	2024-05-09 10:36:06.654202923 +0300
+++ ucode-2023-06-06-c7d84aae/lib/ubus.c	2024-05-09 10:36:06.598203287 +0300
@@ -647,6 +647,9 @@
 	bool prev = uloop_cancelled;
 	bool active;
 
+	if (uloop_fd_set_cb)
+		return true;
+
 	uloop_cancelled = true;
 	active = uloop_cancelling();
 	uloop_cancelled = prev;
